### DISCUSSION TITLE
v5 alpha release and feedback

### DISCUSSION THREAD

**Discussion by mcous** [tracespace/tracespace#412] [Announcements]
Earlier this year, a new alpha version of the long-gestating tracespace version 5 was released: `v5.0.0-alpha.0`! `v5` is a ground-up rewrite of the entire project, with an eye towards building a more sustainable platform moving forward.

We've got a new set of packages to take over from v4. **If you are able to use these alpha versions, we'd love your feedback!** Please add any and all feedback to this discussion thread.

- [@tracespace/parser][] parses Gerber/drill files into abstract syntax trees
  - Replaces gerber-parser
- [@tracespace/plotter][] converts ASTs into image plot trees
  - Replaces gerber-plotter
- [@tracespace/renderer][] renders image plot trees into SVGs
  - Replaces gerber-to-svg
- [@tracespace/identify-layers][] guesses layer types by file name
    - Replaces whats-that-gerber
- [@tracespace/core][] ties everything together to produce full board top/bottom renders
  - Replaces pcb-stackup and pcb-stackup-core
- [@tracespace/cli][] uses @tracespace/core from the command line to parse, plot, and render

[@tracespace/cli]: https://github.com/tracespace/tracespace/tree/v5/packages/cli
[@tracespace/core]: https://github.com/tracespace/tracespace/tree/v5/packages/core
[@tracespace/fixtures]: https://github.com/tracespace/tracespace/tree/v5/packages/fixtures
[@tracespace/identify-layers]: https://github.com/tracespace/tracespace/tree/v5/packages/identify-layers
[@tracespace/parser]: https://github.com/tracespace/tracespace/tree/v5/packages/parser
[@tracespace/plotter]: https://github.com/tracespace/tracespace/tree/v5/packages/plotter
[@tracespace/renderer]: https://github.com/tracespace/tracespace/tree/v5/packages/renderer

Over the coming weeks and months, more alpha versions are going to be released. APIs will probably change drastically, bugs will be fixed, new packages will be released, and the tracespace website will be updated to use the new render pipeline. You will also see the [v5-fixed](https://github.com/tracespace/tracespace/issues?q=is%3Aopen+is%3Aissue+label%3Av5-fixed) label start to appear on issues that have been resolved in v5.

The primary development branch of tracespace/tracespace has switched to the [v5 branch](https://github.com/tracespace/tracespace/tree/v5). The [main branch](https://github.com/tracespace/tracespace/tree/main) still holds the code for v4, and _may_ see bugfix releases - especially if PRs are submitted - but will otherwise be pretty dormant moving forward.

**Comment by shawon1220** [comment]
HI,I have a few questions for @tracespace/core 
1. hyperlink jump error.
![image](https://user-images.githubusercontent.com/10345404/216877233-a54149cd-c626-41a8-9f1c-ceaf1d2112dc.png)

2. index.ts import stringify-svg repeat.

3. when use hast-util-to-html throw error:
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: C:\Users\shawon\ts-koa\node_modules\hast-util-to-html\index.js
require() of ES modules is not supported.
require() of C:\Users\shawon\ts-koa\node_modules\hast-util-to-html\index.js from C:\Users\Administrator\Downloads\ts-koa-starter-master\src\service\home-service.ts is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.

4. renderBoardResult.top  is an instance of Object , then write svg file error-> TypeError [ERR_INVALID_ARG_TYPE]: The "data" argument must be of type string or an instance of Buffer, TypedArray, or DataView

renderBoardResult.top:
`{
  type: 'element',
  tagName: 'svg',
  properties: {
    version: '1.1',
    xmlns: 'http://www.w3.org/2000/svg',
    xmlnsXLink: 'http://www.w3.org/1999/xlink',
    strokeLineCap: 'round',
    strokeLineJoin: 'round',
    strokeWidth: '0',
    fillRule: 'evenodd',
    clipRule: 'evenodd',
    fill: 'currentColor',
    stroke: 'currentColor',
    viewBox: '0 -28.19020000799992 56.5 56.500000007999915'
  },
  children: [
    {
      type: 'element',
      tagName: 'defs',
      properties: {},
      children: [Array]
    },
    {
      type: 'element',
      tagName: 'g',
      properties: [Object],
      children: [Array]
    }
  ]
}`

**Reply by mcous** [reply]
>     1. hyperlink jump error.

Thanks, I fixed this

>     2. index.ts import stringify-svg repeat.

This is incorrect, there is no repeated `import`. There is an `import` followed by an `export`. I have closed #413 accordingly.

>     3. when use hast-util-to-html throw error:
>        Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: C:\Users\shawon\ts-koa\node_modules\hast-util-to-html\index.js
>        require() of ES modules is not supported.
>        require() of C:\Users\shawon\ts-koa\node_modules\hast-util-to-html\index.js from C:\Users\Administrator\Downloads\ts-koa-starter-master\src\service\home-service.ts is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.

This is a build issue with your project that I am unable to help with. `hast-util-to-html` is ESM-only. [Here is a useful writeup](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c) for what that means and what to do.

You can also try using `import {stringifySvg} from '@tracespace/core`

>     4. renderBoardResult.top  is an instance of Object , then write svg file error-> TypeError [ERR_INVALID_ARG_TYPE]: The "data" argument must be of type string or an instance of Buffer, TypedArray, or DataView

Sorry, there is a typo in the @tracespace/core README! The correct code snippet should be:

```js
import fs from 'node:fs/promises'
import {read, plot, renderLayers, renderBoard, stringifySvg} from '@tracespace/core'

const files = [
  'top-copper.gbr',
  'top-solder-mask.gbr',
  'top-silk-screen.gbr',
  'bottom-copper.gbr',
  'bottom-solder-mask.gbr',
  'outline.gbr',
  'drill.xnc',
]

const readResult = await read(files)
const plotResult = plot(readResult)
const renderLayersResult = renderLayers(plotResult)
const renderBoardResult = renderBoard(renderLayersResult)

await Promise.all([
  fs.writeFile('top.svg', stringifySvg(renderBoardResult.top))
  fs.writeFile('bottom.svg', stringifySvg(renderBoardResult.bottom))
])
```

**Comment by shawon1220** [comment]
Hi,Exception when processing larger files.

Error:
RangeError: Maximum call stack size exceeded
      at Object.feed (file:///D:/koa-esm/packages/parser/src/index.ts:67:18)
      at Module.parse (file:///D:/koa-esm/packages/parser/src/index.ts:85:25)
      at file:///D:/koa-esm/packages/core/src/index.ts:92:28
      at Generator.next (<anonymous>)
      at fulfilled (file:///D:/koa-esm/packages/core/src/index.ts:4:58)

Env:
 Node v14.17.3+typescript

File:
[exceeded.zip](https://github.com/tracespace/tracespace/files/10680640/exceeded.zip)

**Reply by mcous** [reply]
Wow this was a fun one. It turns out that the built in `array.push(...sourceArray)` will throw `Maximum call stack size exceeded` if `sourceArray` is too large. See https://es.discourse.group/t/array-method-to-append-another-array-in-place/98

The simple fix is to switch to a for loop. This was a good catch!

**Comment by shawon1220** [comment]
Many svg files generated by pcb exceed 20M or even larger. I use the svgo plug-in to compress it to 9M. Can you consider using this plug-in to optimize the SVG file in v5?  [https://github.com/svg/svgo](url)

**Reply by mcous** [reply]
I'll definitely be trying to optimize the output once I've got more acceptance testing in place. There's definitely a lot of floating point silliness that bloats file sizes.

That being said, I don't think tracespace is going to integrate svgo itself. I'd rather prioritize accuracy file size and let any libray user who wants to pull in `svgo` if needed

**Comment by ElektroLabeu** [comment]
Hi, I have a question. How would I be able to install your script on my hosting? I haven't found anything about it anywhere. Thank you for the answer
